File: android/src/main/java/com/korantom/flutter_mcumgr/ManagerWrapper.java
package com.korantom.flutter_mcumgr;

import android.os.Handler;

import androidx.annotation.NonNull;

import io.flutter.embedding.engine.plugins.FlutterPlugin;
import io.flutter.plugin.common.EventChannel;
import io.flutter.plugin.common.PluginRegistry;


/**
 * ManagerWrapper
 * <p>
 * provides status and progress communication channels
 */
public abstract class ManagerWrapper {
    protected EventChannel progressEventChannel;
    protected EventChannel statusEventChannel;

    protected StreamHandler progressStreamHandler;
    protected StreamHandler statusStreamHandler;

    public ManagerWrapper(String name, @NonNull FlutterPlugin.FlutterPluginBinding flutterPluginBinding) {
        this.progressEventChannel = new EventChannel(flutterPluginBinding.getBinaryMessenger(), String.format("%s/progress", name));
        this.statusEventChannel = new EventChannel(flutterPluginBinding.getBinaryMessenger(), String.format("%s/status", name));

        this.progressStreamHandler = new StreamHandler();
        this.statusStreamHandler = new StreamHandler();


        this.progressEventChannel.setStreamHandler(progressStreamHandler);
        this.statusEventChannel.setStreamHandler(statusStreamHandler);
    }

    public ManagerWrapper(String name, PluginRegistry.Registrar registrar) {
        this.progressEventChannel = new EventChannel(registrar.messenger(), String.format("%s/progress", name));
        this.statusEventChannel = new EventChannel(registrar.messenger(), String.format("%s/status", name));

        this.progressStreamHandler = new StreamHandler();
        this.statusStreamHandler = new StreamHandler();


        this.progressEventChannel.setStreamHandler(progressStreamHandler);
        this.statusEventChannel.setStreamHandler(statusStreamHandler);
    }
}

File: ios/Classes/Model/ManagerWrapper.swift
//
//  ManagerWrapper.swift
//  flutter_mcumgr
//
//  Created by Tomáš Koranda on 06/09/2020.
//

import Foundation
import Flutter

class ManagerWrapper: ProgressStatusChannel{
    let statusEventChannel: FlutterEventChannel
    let progressEventChannel: FlutterEventChannel
    
    let statusStreamHandler: StreamHandler
    let progressStreamHandler: StreamHandler
    
    init(name: String, registrar: FlutterPluginRegistrar) {
        self.statusEventChannel = FlutterEventChannel(name: "\(name)/status", binaryMessenger: registrar.messenger())
        self.progressEventChannel = FlutterEventChannel(name: "\(name)/progress", binaryMessenger: registrar.messenger())
        
        self.statusStreamHandler = StreamHandler()
        self.progressStreamHandler = StreamHandler()
        
        statusEventChannel.setStreamHandler(statusStreamHandler)
        progressEventChannel.setStreamHandler(progressStreamHandler)
    }
    
    enum Status: String {
        case inProgress
        case paused
        case failed
        case success
        case canceled
    }
}
